[NativeAOT-LLVM] Build on Linux

yowl opened the pull request:
This PR reenables the build of the product on Linux, including the LLVM JIT.  Does not build the Wasm runtime or libraries on Linux (to be done in #2569 )

Fixes #1713

yowl committed:
Enable linux builds

yowl committed:
revert package-lock.json

jkotas commented on src/mono/browser/runtime/package-lock.json:
```diff
@@ -1,8 +1,2756 @@
 {
   "name": "@microsoft/dotnet-runtime",
```
Are the changes in this file intentional - why do we need them?

yowl replied:
No, reverted.

yowl commented:
cc @dotnet/nativeaot-llvm

yowl committed:
move NATIVEAOT_CI_WASM_BUILD_EMSDK_PATH to be inline with windows
workflow

yowl committed:
move NATIVEAOT_CI_WASM_BUILD_EMSDK_PATH to be inline with windows

yowl committed:
just use pwsh for installing node.  Delete the .cmd

SingleAccretion commented on eng/native/gen-buildsys.sh:
```diff
@@ -95,6 +95,9 @@ fi
 
 if [[ "$host_arch" == "wasm" ]]; then
     if [[ "$target_os" == "browser" ]]; then
+        if [[ -n "$NATIVEAOT_CI_WASM_BUILD_EMSDK_PATH" ]]; then
+             source $NATIVEAOT_CI_WASM_BUILD_EMSDK_PATH/emsdk_env.sh
+        fi
```
The Windows equivalent is in `eng/common/build.ps1`. Can we do the same with `eng/common/build.sh` to keep the scripts similar?

yowl replied:
Yes, moved.

SingleAccretion commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -206,15 +206,25 @@ jobs:
       - script: call $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-nodejs.cmd $(Build.SourcesDirectory)\wasm-tools
         displayName: Install NodeJS
 
+      # Install Wasm dependencies on Linux: emscripten, LLVM, NodeJS.
+    - ${{ if and(eq(parameters.hostedOs, 'linux'), and(eq(parameters.runtimeFlavor, 'coreclr'), eq(parameters.archType, 'wasm'))) }}:
+      - script: $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-emscripten.sh $(Build.SourcesDirectory)/wasm-tools
+        displayName: Install/activate emscripten
+      - script: pwsh $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-llvm.ps1 -Configs ${{ parameters.buildConfig }} -CI
+        workingDirectory: $(Build.SourcesDirectory)/wasm-tools
+        displayName: Install/build LLVM
+      - script: pwsh $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-nodejs.ps1 $(Build.SourcesDirectory)/wasm-tools
+        displayName: Install NodeJS
```
Would it work to unify the both Windows and Linux on `pwsh`?

yowl replied:
yes, removed the `.cmd`

SingleAccretion commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -206,15 +206,25 @@ jobs:
+    - ${{ if and(eq(parameters.hostedOs, 'linux'), and(eq(parameters.runtimeFlavor, 'coreclr'), eq(parameters.archType, 'wasm'))) }}:
+      - script: $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-emscripten.sh $(Build.SourcesDirectory)/wasm-tools
+        displayName: Install/activate emscripten
+      - script: pwsh $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-llvm.ps1 -Configs ${{ parameters.buildConfig }} -CI
```
`parameters.buildConfig` can be `Checked`, to the normalization of of `Config` in `install-llvm.cmd` needs to be moved into `install-llvm.ps1`.

yowl replied:
Done

SingleAccretion commented on eng/pipelines/runtimelab.yml:
```diff
@@ -59,7 +59,7 @@ extends:
             buildConfig: debug
             platforms:
-#            - linux_x64
+            - linux_x64
```
These `linux_x64` jobs only test upstream code - I think we can do with just one to exercise our changes to the build scripts (say, only Release).

yowl replied:
RIght, have left just Release.

SingleAccretion commented on eng/pipelines/runtimelab/install-emscripten.sh:
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash
```
Perhaps we can use the same powershell sharing strategy as with the the other scripts?

yowl replied:
right, have deletes the .cmd/.sh and just left the `ps1`

SingleAccretion commented on eng/pipelines/runtimelab/install-llvm.ps1:
```diff
@@ -35,23 +41,45 @@ foreach ($Config in $Configs)
     pushd llvm-project
     $BuildDirName = "build-$($Config.ToLower())"
-    mkdir $BuildDirName -Force
+    if ($IsWindows)
+    {
+        mkdir $BuildDirName -Force
+    }
+    else
+    {
+        mkdir $BuildDirName --parents
+    }
```
```suggestion
    New-Item $BuildDirName -Force
```
`mkdir` is a PS alias; not using should avoid the need for the condition.

yowl replied:
Thanks, done with `-ItemType Directory`

SingleAccretion commented on eng/pipelines/runtimelab/install-llvm.ps1:
```diff
@@ -35,23 +41,45 @@ foreach ($Config in $Configs)
     popd
 
-    $CmakeConfigureCommandLine = "-G", "Visual Studio 17 2022", "-DLLVM_INCLUDE_BENCHMARKS=OFF", "-Thost=x64"
+    if ($IsWindows)
+    {
+        $generator="Visual Studio 17 2022"
+    }
+    else
+    {
+        $generator="Unix Makefiles"
+    }
+
+    $CmakeConfigureCommandLine = "-G", "$generator", "-DLLVM_INCLUDE_BENCHMARKS=OFF"
```
```suggestion
    if ($IsWindows)
    {
        $CmakeGenerator = "Visual Studio 17 2022"
    }
    else
    {
        $CmakeGenerator = "Unix Makefiles"
    }

    $CmakeConfigureCommandLine = "-G", "$CmakeGenerator", "-DLLVM_INCLUDE_BENCHMARKS=OFF"
```
Nit: pascal case variables.

yowl replied:
Thanks

SingleAccretion commented on eng/pipelines/runtimelab/install-llvm.ps1:
```diff
@@ -6,6 +6,12 @@ param(
     [switch]$NoBuild
 )
 
+# Set IsWindows if the version of Powershell does not already have it.
+if (!(Test-Path variable:global:IsWindows)) 
+{
+    $IsWindows=[environment]::OSVersion.Platform -eq [PlatformID]::Win32NT
+}
```
```suggestion
# Set IsWindows if the version of Powershell does not already have it.
if (!(Test-Path variable:global:IsWindows)) 
{
    $IsWindows = [Environment]::OSVersion.Platform -eq [PlatformID]::Win32NT
}
```

SingleAccretion commented on eng/pipelines/runtimelab/install-nodejs.ps1:
```diff
@@ -1,7 +1,21 @@
+
+if (!(Test-Path variable:global:IsWindows))
+{
+    $IsWindows=[environment]::OSVersion.Platform -eq [PlatformID]::Win32NT
```
```suggestion
    $IsWindows = [Environment]::OSVersion.Platform -eq [PlatformID]::Win32NT
```

SingleAccretion commented on eng/pipelines/runtimelab/install-nodejs.ps1:
```diff
@@ -33,9 +47,17 @@ if ($RetryCount -le 0)
-Expand-Archive -LiteralPath "$InstallPath\$NodeJSInstallName.zip" -DestinationPath $InstallPath -Force
+if ($IsWindows)
+{
+    Expand-Archive -LiteralPath "$InstallPath\$NodeJSZipName" -DestinationPath $InstallPath -Force
+    $NodeJSExePath = "$InstallPath\$NodeJSInstallName\node.exe"
```
Does `Expand-Archive` not work on Linux (it's a bit surprising)?

yowl replied:
It does, but doesn't support `gz` which is gzip, or `xz` which is some LZMA I think.

yowl replied:
And those are our options for linux https://nodejs.org/dist/v9.9.0/

SingleAccretion commented on src/coreclr/build-runtime.sh:
```diff
@@ -171,6 +171,26 @@ if [[ -n "$__RequestedBuildComponents" ]]; then
 fi
+
+if [[ "$__CMakeTarget" == *"wasmjit"* ]]; then
+  __ExtraCmakeArgs="$__ExtraCmakeArgs -DCLR_CMAKE_BUILD_LLVM_JIT=1"
```
Do we need `__ExtraCmakeArgs` in addition to `__CMakeArgs`?

yowl replied:
missed this one, but is in now.

SingleAccretion commented on src/coreclr/jit/CMakeLists.txt:
```diff
@@ -91,7 +91,9 @@ function(create_standalone_jit)
 
-    find_package(LLVM REQUIRED CONFIG PATHS $ENV{LLVM_CMAKE_CONFIG})
+    # Prefer the version specified in LLVM_CMAKE_CONFIG over the "system" installed version
+    find_package(LLVM REQUIRED CONFIG PATHS $ENV{LLVM_CMAKE_CONFIG} NO_DEFAULT_PATH)
+    find_package(LLVM REQUIRED CONFIG)
```
Does this mean we will search for the system LLVM install too? That would not be correct.

yowl replied:
It does, removed.

SingleAccretion commented on src/native/libs/build-native.sh:
```diff
@@ -60,15 +60,15 @@ source "$__RepoRootDir"/eng/native/build-commons.sh
 # Set cross build
 if [[ "$__TargetOS" == browser ]]; then
-    if [[ -z "$EMSDK_PATH" ]]; then
+    if [[ -z "$EMSDK" ]]; then
```
It would be better to just set `EMSDK_PATH` to `EMSDK` somewhere above s.t. these lines do not need to be modified.

yowl replied:
Done, thanks

SingleAccretion commented on src/tests/Common/tests.targets:
```diff
@@ -6,7 +6,7 @@
 
   <PropertyGroup>
-      <TestAssemblyDir Condition="'$(TestAssemblyDir)' == ''">$(BaseOutputPathWithConfig)\tests\</TestAssemblyDir>
+      <TestAssemblyDir Condition="'$(TestAssemblyDir)' == ''">$(BaseOutputPathWithConfig)\Tests\</TestAssemblyDir>
```
Why does upstream not run into the same problem (in other words - is this an upstream bug)?

yowl replied:
Have reverted this one, it looked wrong, but wasn't the problem I thought it was.

yowl committed:
normalize checked for llvm config

yowl committed:
add missing )

yowl committed:
use powershell only for emscripten install

yowl committed:
use pwsh inline with node

yowl committed:
add ps1 file

yowl committed:
use New-Item correctly

yowl committed:
recommendations for installing llvm

yowl committed:
do not search system paths for LLVM

yowl committed:
feedback tidy

yowl committed:
do not use __ExtraCmakeArgs

yowl committed:
feedback, tidy/consolidate

yowl committed:
Merge branch 'linux' of github.com:yowl/runtimelab into linux

yowl committed:
revert to complex directory creation

SingleAccretion commented on eng/common/build.sh:
```diff
@@ -87,6 +87,8 @@ prepare_machine=false
 verbosity='minimal'
 runtime_source_feed=''
 runtime_source_feed_key=''
+host_arch=''
+target_os=''
```
Are these only required for `NATIVEAOT_CI_WASM_BUILD_EMSDK_PATH`? It's not clear what they're for.

yowl replied:
Its to avoid 
```
/home/scott/github/runtimelab/eng/common/build.sh: line 245: host_arch: unbound variable
```
`${host_arch-''}` might be another way to solve it, or make it a mandatory parameter.

SingleAccretion replied:
But the only uses of `host_arch` that I see look to be new (added in this change). My suggestion is delete them - or are they needed for something?

SingleAccretion commented on eng/pipelines/runtimelab/install-emscripten.ps1:
```diff
@@ -0,0 +1,23 @@
+
+python ./emsdk.py install 3.1.47
+
+./emsdk activate 3.1.47
```
We have `install-emscripten.cmd` referenced in `docs/using-nativeaot/prerequisites.md` and also `docs/workflow/building/coreclr/nativeaot.md`, these need to be updated now.

Also, it is odd that the first line calls `.py` directly while the second goes through the wrapper script.

yowl commented:
Lots of little commits, but think that's all the comments above addressed.

SingleAccretion commented on src/coreclr/build-runtime.sh:
```diff
@@ -185,7 +205,7 @@ fi
 eval "$__RepoRootDir/eng/native/version/copy_version_files.sh"
 
-build_native "$__HostOS" "$__HostArch" "$__ProjectRoot" "$__IntermediatesDir" "$__CMakeTarget" "$__CMakeArgs" "CoreCLR component"
+build_native "$__HostOS" "$__HostArch" "$__ProjectRoot" "$__IntermediatesDir" "$__CMakeTarget" "$__CMakeArgs $__ExtraCmakeArgs" "CoreCLR component"
```
```suggestion
build_native "$__HostOS" "$__HostArch" "$__ProjectRoot" "$__IntermediatesDir" "$__CMakeTarget" "$__CMakeArgs" "CoreCLR component"
```
Now removed.

SingleAccretion commented on src/coreclr/jit/CMakeLists.txt:
```diff
@@ -91,7 +91,8 @@ function(create_standalone_jit)
     set_target_properties(${TARGETDETAILS_TARGET} PROPERTIES CLR_EH_CONTINUATION OFF)
 
-    find_package(LLVM REQUIRED CONFIG PATHS $ENV{LLVM_CMAKE_CONFIG})
+    # Prefer the version specified in LLVM_CMAKE_CONFIG over the "system" installed version
+    find_package(LLVM REQUIRED CONFIG PATHS $ENV{LLVM_CMAKE_CONFIG} NO_DEFAULT_PATH)
```
```suggestion
    # Only look at "our" LLVM with LLVM_CMAKE_CONFIG.
    find_package(LLVM REQUIRED CONFIG PATHS $ENV{LLVM_CMAKE_CONFIG} NO_DEFAULT_PATH)
```

SingleAccretion commented on eng/pipelines/runtimelab/install-emscripten.ps1:
```diff
@@ -0,0 +1,23 @@
+    [string]$InstallDir
+)
+
+New-Item -ItemType Directory -Force -ErrorAction SilentlyContinue -Path (Split-Path -Path $InstallDir -Parent) -Name (Split-Path -Path $InstallDir -Leaf)
```
Does this really need to be this elaborate?

I just tried:
```
/home/singleaccretion> New-Item /home/singleaccretion/test/xyz/dir -Force -ItemType Directory
```
And it created the nested directory as expected.

yowl replied:
Its ok on Linux, but with a drive letter it fails.  We could strip the drive letter and assume we are "in" that drive, but it probably isn't much easier to read.
```
PS C:\Users\scott> new-item -ItemType Directory -Force -Name c:/tmp/1/2
new-item : The given path's format is not supported.
At line:1 char:1
+ new-item -ItemType Directory -Force -Name c:/tmp/1/2
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [New-Item], NotSupportedException
    + FullyQualifiedErrorId : System.NotSupportedException,Microsoft.PowerShell.Commands.NewItemCommand

PS C:\Users\scott> new-item -ItemType Directory -Force -Name /tmp/1/2


    Directory: C:\Users\scott\tmp\1


Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d-----        11/05/2024     16:54                2
```

SingleAccretion replied:
> `new-item -ItemType Directory -Force -Name c:/tmp/1/2`

I think this should use `-Path`. I tried:
```
PS C:\Users\Accretion\AppData\Local\Temp> new-item -Force -ItemKind Directory -Path C:\Users\Accretion\AppData\Local\Temp/xyz/rot/tot

    Directory: C:\Users\Accretion\AppData\Local\Temp\xyz\rot

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
d----          13.05.2024    00:22                tot
```

SingleAccretion commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -206,13 +206,23 @@ jobs:
+        displayName: Install NodeJS
+
+      # Install Wasm dependencies on Linux: emscripten, LLVM, NodeJS.
+    - ${{ if and(eq(parameters.hostedOs, 'linux'), and(eq(parameters.runtimeFlavor, 'coreclr'), eq(parameters.archType, 'wasm'))) }}:
```
Can we collapse the Windows/Linux into one section now (will require putting Emscripten on the same `pwsh` plan)?

yowl replied:
Yes, done

yowl committed:
Update src/coreclr/build-runtime.sh

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
Update src/coreclr/jit/CMakeLists.txt

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

SingleAccretion commented on eng/pipelines/runtimelab/install-emscripten.ps1:
```diff
@@ -0,0 +1,23 @@
+    [string]$InstallDir
+)
+
+New-Item -ItemType Directory -Force -ErrorAction SilentlyContinue -Name $InstallDir
+
+$ErrorActionPreference="Stop"
+
+Set-Location -Path $InstallDir
```
```suggestion
$ErrorActionPreference="Stop"

New-Item -Path $InstallDir -ItemType Directory -Force
Set-Location -Path $InstallDir
```
I think this ought to work...

yowl replied:
Yes, indeed that does.

SingleAccretion commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -206,13 +206,13 @@ jobs:
+      - script: pwsh $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-emscripten.ps1 $(Build.SourcesDirectory)/wasm-tools
         displayName: Install/activate emscripten
-      - script: call $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-llvm.cmd $(Build.SourcesDirectory)\wasm-tools $(Build.SourcesDirectory) ${{ parameters.buildConfig }}
+      - script: pwsh $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-llvm.ps1 $(Build.SourcesDirectory)\wasm-tools $(Build.SourcesDirectory) ${{ parameters.buildConfig }}
```
```suggestion
      - script: pwsh $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-llvm.ps1 $(Build.SourcesDirectory)/wasm-tools $(Build.SourcesDirectory) -Configs ${{ parameters.buildConfig }} -CI
```
The cmake detection and `cd $(Build.SourcesDirectory)/wasm-tools` need to be migrated from `install-llvm.cmd` to `install-llvm.ps1`.

yowl replied:
The cmake detection, i.e. executing `init-vs-env.cmd` as we do, doesn't seem necessary, and puts the `CMakePath` at the end anyway so doesn't precede anything already in the path.  It is possible to do from powershell, but would involve a wrapper `cmd` to `echo` the `CMakePath` back to the powershell do you scoping of environment variables and the lack or an equivalent `call` in powershell.  Question is, do we really need it?

SingleAccretion replied:
I had assumed it was needed. If that's not the case - great!

yowl replied:
We do need `cmake.exe` in the path, so have added this back, which means we need a `cmd` for the windows side.

yowl commented:
"But the only uses of host_arch that I see look to be new (added in this change). My suggestion is delete them - or are they needed for something?"

You are right, I don't actually need it here, but I will for `./build.sh clr.aot -c Debug -a wasm -os browser`

SingleAccretion commented:
My preference would be introduce them in the WASM change. In general, it looks unexpected to me that we need to change `eng/common/build.sh`, since these scripts are used upstream for `-a wasm -os browser` builds too.

yowl committed:
Ensure Powershell on linux

Remove wasm build options
add cmake to path on windows
Use more Powershell where possible
Simplify New-Item use
Feedback

yowl commented:
> My preference would be introduce them in the WASM change. In general, it looks unexpected to me that we need to change `eng/common/build.sh`, since these scripts are used upstream for `-a wasm -os browser` builds too.

OK, removed here.

yowl commented:
Was on a break last week, so did this with lots of trail and error, but have now rebased to remove that noise and is green.

SingleAccretion commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -222,10 +224,18 @@ jobs:
-      # Install LLVM for the win-x64 build as it will build the clrjit for browser_wasm cross compilation
-      - script: call $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-llvm.cmd $(Build.SourcesDirectory)\wasm-tools $(Build.SourcesDirectory) ${{ parameters.buildConfig }}
-        displayName: Install/build LLVM
+    - ${{ if eq(parameters.archType, 'x64') }}:
+      - ${{ if eq(parameters.osGroup, 'windows') }}:
+        - script: call $(Build.SourcesDirectory)/eng/pipelines/runtimelab/set-cmake-path.cmd
+          displayName: Set CMake path
+        - script: pwsh $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-llvm.ps1 -CI -InstallDir $(Build.SourcesDirectory)/wasm-tools -Configs ${{ parameters.buildConfig }}
+          displayName: Install/build LLVM
+      - ${{ if eq(parameters.osGroup, 'linux') }}:
+        - script: $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-pwsh.sh $(Build.SourcesDirectory)/wasm-tools
+          displayName: Install Powershell 7
+        # Install LLVM for the win-x64 and linux-x64 build as it will build the clrjit for browser_wasm cross compilation
+        - script: $(Build.SourcesDirectory)/wasm-tools/powershell7/pwsh $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-llvm.ps1 -CI -InstallDir $(Build.SourcesDirectory)/wasm-tools -Configs ${{ parameters.buildConfig }}
+          displayName: Install/build LLVM
```
This will install tools for host configurations (`windows_x64`, `linux_x64`) that do not actually target WASM. I think the LLVM stuff should be deleted for `windows_x64` and `Install Powershell 7` folded into the above block (under `{{ if and(eq(parameters.runtimeFlavor, 'coreclr'), eq(parameters.archType, 'wasm')) }}:`) under `if ne(parameters.hostedOs, 'windows')` (this way it will, hopefully, also work for macOS if/when that comes).

yowl replied:
Thanks, done.  The Linux build isn't pulling in these tools with this change but that's fine, they will be with #2569

SingleAccretion commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -222,10 +224,18 @@ jobs:
+        - script: $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-pwsh.sh $(Build.SourcesDirectory)/wasm-tools
+          displayName: Install Powershell 7
+        # Install LLVM for the win-x64 and linux-x64 build as it will build the clrjit for browser_wasm cross compilation
+        - script: $(Build.SourcesDirectory)/wasm-tools/powershell7/pwsh $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-llvm.ps1 -CI -InstallDir $(Build.SourcesDirectory)/wasm-tools -Configs ${{ parameters.buildConfig }}
```
`install-pwsh.sh` sets `PATH`. Does a plain `pwsh $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-llvm.ps1 ...` not work?

yowl replied:
I think that will work, yes.  Its not actually exercised with this PR but should be fine.

yowl committed:
feedback, move tool installations to correct place.

SingleAccretion commented on eng/pipelines/common/global-build-job.yml:
```diff
@@ -208,11 +208,17 @@ jobs:
+      - script: pwsh $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-emscripten.ps1 $(Build.SourcesDirectory)/wasm-tools
         displayName: Install/activate emscripten
-      - script: call $(Build.SourcesDirectory)/eng/pipelines/runtimelab/install-llvm.cmd $(Build.SourcesDirectory)\wasm-tools $(Build.SourcesDirectory) ${{ parameters.buildConfig }}
+      - script: call $(Build.SourcesDirectory)/eng/pipelines/runtimelab/set-cmake-path.cmd
```
I suspect this will need to be put under `if eq(parameters.hostedOs, 'windows')`, but let's leave that for the next change.

SingleAccretion reviewed (approved):
LGTM, thank you!

jkotas reviewed (approved):
Thank you!

sardonic1 commented:
What's the state of affairs here?  I'd like to be able to automate wasi-wasm compilations using Linux (docker on Mac).  I see that @yowl has done a lot of work here, but I'm unsure how to leverage it if it is indeed completed.  For example, I can't see any artifacts published like runtime.linux-x64.Microsoft.DotNet.ILCompiler.LLVM.

SingleAccretion commented:
@sardonic1 there is still one more task left here, which is to enable the official build, which produces the packages.

sardonic1 commented:
Thanks for that.  Is there a rough idea of when this feature might be available?  Any idea if any of this experimental branch will make it to a .NET 9 release?

SingleAccretion commented:
> Thanks for that. Is there a rough idea of when this feature might be available?

I think we can make the official build work pretty soon ('few days' soon). It is not a large change.

> Any idea if any of this experimental branch will make it to a .NET 9 release?

The compiler and runtime - no. Note: .NET 9 is pretty much "done" at this point - upstream `main` will soon switch to .NET 10. A certain amount of recent WASI-related libraries work will be in .NET 9.
